source/fuzz/fuzzer_pass.cpp

```diff
@@ -595,5 +597,99 @@ opt::BasicBlock* FuzzerPass::GetOrCreateSimpleLoopPreheader(
   return &*function->FindBlock(preheader_id);
 }
 
+uint32_t FuzzerPass::FindOrCreateLocalVariable(
+    uint32_t pointer_type_id, uint32_t function_id,
+    bool pointee_value_is_irrelevant) {
+  auto pointer_type = GetIRContext()->get_type_mgr()->GetType(pointer_type_id);
+  // No unused variables in release mode.
+  (void)pointer_type;
+  assert(pointer_type && pointer_type->AsPointer() &&
+         pointer_type->AsPointer()->storage_class() ==
+             SpvStorageClassFunction &&
+         "The pointer_type_id must refer to a defined pointer type with "
+         "storage class Function");
+  auto function = fuzzerutil::FindFunction(GetIRContext(), function_id);
+  assert(function && "The function must be defined.");
+
+  // All of the local variable declarations are located in the first block.
+  auto block = function->begin();
+  for (auto& instruction : *block) {
+    if (instruction.opcode() != SpvOpVariable) {
+      continue;
+    }
+    // Every OpVariable has its type id defined.
+    if (instruction.type_id() != pointer_type_id) {
+      continue;
+    }
+    // Check if the found variable is marked with PointeeValueIsIrrelevant
+    // according to |pointee_value_is_irrelevant|.
+    if (GetTransformationContext()->GetFactManager()->PointeeValueIsIrrelevant(
+            instruction.result_id()) != pointee_value_is_irrelevant) {
+      continue;
+    }
+    return instruction.result_id();
+  }
+
+  // No such variable was found. Apply a transformation to get one.
+  uint32_t pointee_type_id = fuzzerutil::GetPointeeTypeIdFromPointerType(
+      GetIRContext(), pointer_type_id);
+  uint32_t result_id = GetFuzzerContext()->GetFreshId();
+  ApplyTransformation(TransformationAddLocalVariable(
+      result_id, pointer_type_id, function_id,
+      FindOrCreateZeroConstant(pointee_type_id, true),
+      pointee_value_is_irrelevant));
+  return result_id;
+}
+
+uint32_t FuzzerPass::FindOrCreateGlobalVariable(
+    uint32_t pointer_type_id, bool pointee_value_is_irrelevant) {
+  auto pointer_type = GetIRContext()->get_type_mgr()->GetType(pointer_type_id);
+  // No unused variables in release mode.
+  (void)pointer_type;
+  assert(
+      pointer_type && pointer_type->AsPointer() &&
+      (pointer_type->AsPointer()->storage_class() == SpvStorageClassPrivate ||
+       pointer_type->AsPointer()->storage_class() ==
+           SpvStorageClassWorkgroup) &&
+      "The pointer_type_id must refer to a defined pointer type with storage "
+      "class Private or Workgroup");
+  for (auto& instruction : GetIRContext()->module()->types_values()) {
+    if (instruction.opcode() != SpvOpVariable) {
+      continue;
+    }
+    // Every OpVariable has its type id defined.
+    if (instruction.type_id() != pointer_type_id) {
+      continue;
+    }
```

Comment:
It sounds like this should be an assert?

Reply:
I don't think so, because there could many local/global variables that have instruction.type_id() different than the supplied pointer_type_id. In that case we just continue to the next instruction.

Reply:
Ah sorry you are right! I think the comment confused me. I somehow just assumed the code was checking that the type id was non-zero.

Maybe consider changing the comment to: 

`The existing OpVariable must have type |pointer_type_id|.`

You could also add a comment at the start of the loop (perhaps before your "All of the local variable..." comment) saying something like: "First, we try to find a suitable existing variable."